### PR TITLE
FP-0010-ngrx-GETALL

### PR DESCRIPTION
**Pull request by TesaRocks**
Hi Diego, please have a look at the GETALL implementation through NGRX-
I had to change my tsconfig file with
strictnullkchecks set to FALSE in order to force it to compile.

Can you please give me a hand on how to fix this?

Thanks

### CODE REVIEW

File: src/app/users/store/users.actions.ts [hunk 1]
```diff
@@ -3,16 +3,21 @@ import { IUser } from '../user.interface';
 
 const begin = createAction('[Users] Get All - Begin');
 const success = createAction(
-  '[Users] Get All - Success',
+  '[Users Effect] Get All - Success',
```
Comment (diegolaciar) [inline src/app/users/store/users.actions.ts]: Does not get sense form me this name 
[Users Effect]

The idea of this naming convention is to set the name of the module there.
[Users]

File: src/app/users/store/users.actions.ts [hunk 2]
```diff
@@ -3,16 +3,21 @@ import { IUser } from '../user.interface';
...
   props<{ users: IUser[] }>()
 );
-const failure = createAction(
-  '[Users] Get All - Failure',
-  props<{ error: any }>()
+
+//delete user
+const deleteUserBegin = createAction(
+  '[Users] Delete - Begin',
```
Comment (diegolaciar) [inline src/app/users/store/users.actions.ts]: Here you made it right,  Always check that you should be consistent on naming conventions.

File: src/app/users/store/users.actions.ts [hunk 3]
```diff
@@ -3,16 +3,21 @@ import { IUser } from '../user.interface';
...
+  props<{ id: number }>()
 );
 
+const deleteUserSuccess = createAction('[Users Effect] Delete - Success');
+
 export const getAllUsersActions = {
```
Comment (diegolaciar) [inline src/app/users/store/users.actions.ts]: this is very wrong. does not have ANY sense this way of organizing your actions.
export const getAllUsersActions   ->  OK you put the actions for getAll, why you also add there actions for delete ?
YOu just need to add a new export with the same pattern.  eg:

```typescript
export const getAllUsersActions = {
begin,
 success,
 failure
}

export const deleteUserActions = {
 begin,
 success,
 failure
}

export const createUserActions = {
 begin,
 success,
 failure
};

export const updateUserActions = {
 begin,
 success,
 failure
}

export const getUserActions = {
 begin,
 success,
 failure
}

export const updateUserActions = {
 begin,
 success,
 failure
}

```

File: src/app/users/store/users.effects.ts [hunk 4]
```diff
@@ -1,8 +1,15 @@
 import { Injectable } from '@angular/core';
```
Comment (diegolaciar) [inline src/app/users/store/users.effects.ts:1]: users.effects.ts   

This is starting to look horrible. Check this:
YOu are stating to make a complex app.  If you put all the code in one file, that file will soon or later grow and look complex for follow. You are putting all your cats on the same bag mate.


You should try to use a file per effect, and a file per reducer.  and try to follow some kind of naming conventions for everything you write.

instead of users.effects.ts      

at least  for all your async (with effect code)  NGRX code
```typescript
getAll.users.effect.ts   
getAll.users.reducer.ts  
get.users.effect.ts   
get.users.reducer.ts  
create.users.effect.ts   
create.users.reducer.ts   
update.users.effect.ts   
update.users.reducer.ts   
delete.users.effect.ts   
delete.users.reducer.ts   
```

File: src/app/users/store/users.effects.ts [hunk 5]
```diff
@@ -1,8 +1,15 @@
...
+  mergeMap,
+  catchError,
+  concatMap,
+  exhaustMap,
+} from 'rxjs/operators';
 import { UserService } from '../user.service';
+import { getAllUsersActions } from './users.actions';
 
 @Injectable()
 export class UserEffects {
```
Comment (diegolaciar) [inline src/app/users/store/users.effects.ts]: this looks like a mixture of stuff, this is maybe because of the lack of code distribution among files with a clear meaning.

File: src/app/users/store/users.reducer.ts [hunk 6]
```diff
@@ -1,41 +1,51 @@
 import { IUser } from '../user.interface';
 import { createReducer, on, Action } from '@ngrx/store';
 import { getAllUsersActions } from './users.actions';
+import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';
+import { state } from '@angular/animations';
 
+//export interface IUsersInitialState extends EntityState<IUser> {
 export interface IUsersInitialState {
   users: IUser[];
-  getAllUsersPending: boolean;
```
Comment (diegolaciar) [inline src/app/users/store/users.reducer.ts:7]: and why did you remove this?

File: src/app/users/store/users.reducer.ts [hunk 7]
```diff
@@ -1,41 +1,51 @@
...
+
+//users: [],
+//  selectedUserId: null,
+//});
 
 const userReducer = createReducer(
   usersInitialState,
   on(getAllUsersActions.begin, (state) => ({
     ...state,
-    getAllUsersPending: true,
```
Comment (diegolaciar) [inline src/app/users/store/users.reducer.ts:20]: why did you remove all this? This is de essence of async functionality. With this, you can show errors of activity indication to the end-users of your app!

File: tsconfig.json [hunk 8]
```diff
@@ -6,6 +6,7 @@
     "outDir": "./dist/out-tsc",
     "forceConsistentCasingInFileNames": true,
     "strict": true,
+    "strictNullChecks":false,
```
Comment (diegolaciar) [inline tsconfig.json]: Why mate? This is not the solution. This is like put the trash under the carpet of your living. Strict null checks makes your code a lot more robust. 
Doing this you are taking an architectural decision, basically, you are setting the quality bar of your project a step-down.

As this is a whiteboard-clean project from the beginning, you must keep the quality where it was. 
Set this setting to true and fix the issues, please.

**Review by diegolaciar — CHANGES_REQUESTED** [review]
Hi @TesaRocks,

I made many comments on the code.  This probable it will look complex probable we should do a call to clarifying some of the things there.

As you know I'm a little busy right now.

Please, take a closer look at all the comments, check what makes sense for you to fix. Then let's do a call.

Cheers mate,
Diego

File: src/app/users/store/users.actions.ts (hunk 1 above)
Comment (TesaRocks) [reply src/app/users/store/users.actions.ts]: Thanks Diego, I was following a naming convention seen on a video. Changed right now

File: src/app/users/store/users.actions.ts (hunk 2 above)
Comment (TesaRocks) [reply src/app/users/store/users.actions.ts]: done!!

File: src/app/users/store/users.actions.ts (hunk 3 above)
Comment (TesaRocks) [reply src/app/users/store/users.actions.ts]: done thanks!!!

Comment (TesaRocks) [reply src/app/users/store/users.actions.ts]: I cannot name begin, success and failure more than once.. How can I solve this?

File: src/app/users/store/users.effects.ts (hunk 5 above)
Comment (TesaRocks) [reply src/app/users/store/users.effects.ts]: fixed thanks

File: src/app/users/store/users.reducer.ts (hunk 6 above)
Comment (TesaRocks) [reply src/app/users/store/users.reducer.ts:7]: I cannot use the same interface for the rest of the reducers because it will have an unproper attribute. 
GetAllUsers is only applicable for the get all users action.
Please let me know.

File: src/app/users/store/users.reducer.ts (hunk 7 above)
Comment (TesaRocks) [reply src/app/users/store/users.reducer.ts:20]: I replaced it with an error message,. please let me know if you agree

File: tsconfig.json (hunk 8 above)
Comment (TesaRocks) [reply tsconfig.json]: I am really struggling to fix this issue, I would appreciate a help from you. Thansk!

File: src/app/users/store/users.reducer.ts (hunk 6 above)
Comment (diegolaciar) [reply src/app/users/store/users.reducer.ts:7]: Wrong concept,  you must use the same interface of users Feature initial state.
this pattern should be used for this interface.
```getAllUsersPending
getAllUserseError
updateUsersPending
updateUsersError
getUsersPending
getUsersError
createUsersPending
createUsersError
```

File: src/app/users/store/users.reducer.ts (hunk 7 above)
Comment (diegolaciar) [reply src/app/users/store/users.reducer.ts:20]: OK you you dont want to use de store for the error (but is a good practice to program that).  but de pending flag is a must that have to be there.

File: tsconfig.json (hunk 8 above)
Comment (diegolaciar) [reply tsconfig.json]: OK, what's the problem?  You must set this to true, then build and fix each strict comparison (null check) you have on the code. the compiler should tell you the file and line where the error occurs?

File: src/app/users/store/users.actions.ts (hunk 3 above)
Comment (diegolaciar) [reply src/app/users/store/users.actions.ts]: IF you are doing it all on the same file (that is not the idea as I mentioned) you can fix that easy like this.

```typescript
export const updateUserActions = {
 begin: updateBeging,
 success: updateSuccess,
 failure: updateFailure
}


export const updateUserActions = {
 begin,
 success,
 failure
}```

File: src/app/users/store/users.reducer.ts (hunk 6 above)
Comment (TesaRocks) [reply src/app/users/store/users.reducer.ts:7]: I did an aproch similar to a video that I have been following. Please let me know If it is ok, otherwise I change it as you say. Thanks

File: src/app/users/store/users.effects.ts (hunk 4 above)
Comment (TesaRocks) [reply src/app/users/store/users.effects.ts:1]: Before I do this, please have a look at my file now.. I finihsed with the crud implemetation via ngrx

File: src/app/users/store/users.actions.ts (hunk 3 above)
Comment (TesaRocks) [reply src/app/users/store/users.actions.ts]: Ok I did it in a similar way, before reading this message, please let me know
Thanks
